index.js

```diff
@@ -46,14 +51,18 @@ function sanitize(input, replacement) {
     .replace(reservedRe, replacement)
     .replace(windowsReservedRe, replacement)
     .replace(windowsTrailingRe, replacement);
-  return truncate(sanitized, 255);
+  return preserveFileExt ? 
+    truncate(sanitized, Math.max(0, truncateLength - extname(sanitized).length)) + truncate(extname(sanitized), truncateLength)
+    : truncate(sanitized, truncateLength);
```

Comment:
Nit: This ternary is a bit hairy. An if/else with some intermediate variables might make it more readable.

```js
if (preserveFileExt) {
  var maxBaseLength = Math.max(0, truncateLength - extname(sanitized).length);
  var truncatedBase = truncate(sanitized, maxBaseLength);
  var truncatedExt = truncate(extname(sanitized), truncateLength);
  return truncatedBase + truncatedExt;
} else {
  return truncate(sanitized, truncateLength);
}
```